src/parser/color.rs

```diff
@@ -0,0 +1,641 @@
+// <rgb> ::= "rgb(" <0-255> "," <0-255> "," <0-255> ")" | "rgba(" <0-255> "," <0-255> "," <0-255> "," <alpha> ")"
+// <hex> ::= "#" <hex-byte> <hex-byte> <hex-byte> | "#" <hex-byte> <hex-byte> <hex-byte> <hex-byte> | "#" <hex-digit> <hex-digit> <hex-digit> | "#" <hex-digit> <hex-digit> <hex-digit> <hex-digit>
+// <hex-digit> ::= <digit> | "a" - "f" | "A" - "F"
+// <hex-byte> ::= <hex-digit> <hex-digit>
+// <hsl> ::= "hsl(" <0-360> "," <0-100> "%," <0-100> "%)" | "hsla(" <0-360> "," <0-100> "%," <0-100> "%," <alpha> ")"
+// <alpha> ::= "0." <digits> | "." <digits> | 1 | 0
+
+use super::*;
+use crate::tokenizer::*;
+
+#[derive(Debug, PartialEq)]
+pub enum Color {
+    Black,
+    Silver,
+    Gray,
+    Grey,
+    White,
+    Maroon,
+    Red,
+    Purple,
+    Fuchsia,
+    Green,
+    Lime,
+    Olive,
+    Yellow,
+    Navy,
+    Blue,
+    Teal,
+    Aqua,
+    Rgb { r: f64, g: f64, b: f64, a: f64 },
+    Hsl { h: f64, s: f64, l: f64, a: f64 },
+}
+
+fn parse_percent<I: Iterator<Item = char>>(
+    parser: &mut Parser<I>,
+    min: f64,
+    max: f64,
+) -> Result<f64, ParsingError> {
+    match parser.tokens.next() {
+        Some(token_at) => match token_at.token {
+            Token::Percentage(val) => {
+                if val >= min && val <= max {
+                    Ok(val)
+                } else {
+                    Err(ParsingError::wrong_token(
+                        token_at,
+                        &format!("a percentage between {min}% and {max}%"),
+                    ))
+                }
+            }
+            _ => Err(ParsingError::wrong_token(token_at, "a percentage")),
+        },
+
+        None => Err(ParsingError::end_of_file("a percentage")),
+    }
+}
+
+fn parse_num<I: Iterator<Item = char>>(
+    parser: &mut Parser<I>,
+    min: f64,
+    max: f64,
+) -> Result<f64, ParsingError> {
+    match parser.tokens.next() {
+        Some(token_at) => match token_at.token {
+            Token::Number(val) => {
+                if val >= min && val <= max {
+                    Ok(val)
+                } else {
+                    Err(ParsingError::wrong_token(
+                        token_at,
+                        &format!("a number between {min} and {max}"),
+                    ))
+                }
+            }
+            _ => Err(ParsingError::wrong_token(token_at, "a number")),
+        },
+
+        None => Err(ParsingError::end_of_file("a number")),
+    }
+}
+
+impl Parsable for Color {
+    fn parse<I: Iterator<Item = char>>(parser: &mut Parser<I>) -> Result<Self, ParsingError> {
+        match parser.tokens.next() {
+            Some(token_at) => match &token_at.token {
+                Token::Identifier(key) => match key.as_str() {
+                    "black" => Ok(Color::Black),
+                    "silver" => Ok(Color::Silver),
+                    "gray" => Ok(Color::Gray),
+                    "grey" => Ok(Color::Grey),
```

Comment:
If we wanted we could make both of these strings go to the same enum variant but this is ok too